resolves #2666 replace glob by picomatch

ggrossetie opened the pull request.

ggrossetie commented:
Windows is always a bit tricky but I guess glob expressions should use `/` (and not `\\`). Cannot find a proper reference but even .NET is using `/`: https://learn.microsoft.com/en-us/dotnet/core/extensions/file-globbing

Anyway, `picomatch` with the `windows` option will still match Windows paths even if the glob expression is using `/`.

ggrossetie commented:
Regarding dependencies, I don't remember why `node_modules` is commited but as you can see `picomatch` removes 11 dependencies:

- balanced-match
- brace-expansion
- concat-map
- fs.realpath
- glob
- inflight
- inherits
- minimatch
- once
- path-is-absolute
- wrappy

takaram commented:
I did some experiment with this branch, and found some problems.

### 1. `Dir.glob` returns an empty array when there are directories the user cannot access
A simple pattern like `/dir/*` does not work if `/dir` contains a directory which the user does not have permission to access.
It works with current master branch.

```console
$ ls -l /tmp/glob_test_2
total 4
-rw-r--r-- 1 takaram takaram    0 Aug  7 20:45 bar
drwx------ 2 root    root    4096 Aug  7 20:45 dir_not_accesible
-rw-r--r-- 1 takaram takaram    0 Aug  7 20:45 foo
$ whoami
takaram
$ bin/opal -r nodejs/dir -e 'p Dir.glob("/tmp/glob_test_2/*")'
[]
```

### 2. Performance regression
`Dir.glob` takes more time than before with a directory which has many subdirectories.

```console
$ echo /tmp/glob_test/{1..100}/{1..100}/{1..30} | xargs mkdir -p
$ time bin/opal -r nodejs/dir -e 'Dir.glob("/tmp/glob_test/*")'
```

- Before: 0.463s
- After: 2.045s

ggrossetie commented:
Thanks for your input! I will take a look 👀

ggrossetie commented:
> A simple pattern like /dir/* does not work if /dir contains a directory which the user does not have permission to access.
It works with current master branch.

The reason is that `fs.readdirSync(base, {recursive: true})` stops at the first error which is a bit unfortunate.
I guess we will need to provide our own recursive function to list files in directory.

> Dir.glob takes more time than before with a directory which has many subdirectories.

We can optimize it when the glob expression is not recursive

ggrossetie commented:
I did some changes, could you please do some additional experiments? 🥺 

#### Before

```
$ time bin/opal -r nodejs/dir -e 'p Dir.glob("/tmp/glob_test/*")'
["/tmp/glob_test/1", "/tmp/glob_test/10", "/tmp/glob_test/100", "/tmp/glob_test/11", "/tmp/glob_test/12", "/tmp/glob_test/13", "/tmp/glob_test/14", "/tmp/glob_test/15", "/tmp/glob_test/16", "/tmp/glob_test/17", "/tmp/glob_test/18", "/tmp/glob_test/19", "/tmp/glob_test/2", "/tmp/glob_test/20", "/tmp/glob_test/21", "/tmp/glob_test/22", "/tmp/glob_test/23", "/tmp/glob_test/24", "/tmp/glob_test/25", "/tmp/glob_test/26", "/tmp/glob_test/27", "/tmp/glob_test/28", "/tmp/glob_test/29", "/tmp/glob_test/3", "/tmp/glob_test/30", "/tmp/glob_test/31", "/tmp/glob_test/32", "/tmp/glob_test/33", "/tmp/glob_test/34", "/tmp/glob_test/35", "/tmp/glob_test/36", "/tmp/glob_test/37", "/tmp/glob_test/38", "/tmp/glob_test/39", "/tmp/glob_test/4", "/tmp/glob_test/40", "/tmp/glob_test/41", "/tmp/glob_test/42", "/tmp/glob_test/43", "/tmp/glob_test/44", "/tmp/glob_test/45", "/tmp/glob_test/46", "/tmp/glob_test/47", "/tmp/glob_test/48", "/tmp/glob_test/49", "/tmp/glob_test/5", "/tmp/glob_test/50", "/tmp/glob_test/51", "/tmp/glob_test/52", "/tmp/glob_test/53", "/tmp/glob_test/54", "/tmp/glob_test/55", "/tmp/glob_test/56", "/tmp/glob_test/57", "/tmp/glob_test/58", "/tmp/glob_test/59", "/tmp/glob_test/6", "/tmp/glob_test/60", "/tmp/glob_test/61", "/tmp/glob_test/62", "/tmp/glob_test/63", "/tmp/glob_test/64", "/tmp/glob_test/65", "/tmp/glob_test/66", "/tmp/glob_test/67", "/tmp/glob_test/68", "/tmp/glob_test/69", "/tmp/glob_test/7", "/tmp/glob_test/70", "/tmp/glob_test/71", "/tmp/glob_test/72", "/tmp/glob_test/73", "/tmp/glob_test/74", "/tmp/glob_test/75", "/tmp/glob_test/76", "/tmp/glob_test/77", "/tmp/glob_test/78", "/tmp/glob_test/79", "/tmp/glob_test/8", "/tmp/glob_test/80", "/tmp/glob_test/81", "/tmp/glob_test/82", "/tmp/glob_test/83", "/tmp/glob_test/84", "/tmp/glob_test/85", "/tmp/glob_test/86", "/tmp/glob_test/87", "/tmp/glob_test/88", "/tmp/glob_test/89", "/tmp/glob_test/9", "/tmp/glob_test/90", "/tmp/glob_test/91", "/tmp/glob_test/92", "/tmp/glob_test/93", "/tmp/glob_test/94", "/tmp/glob_test/95", "/tmp/glob_test/96", "/tmp/glob_test/97", "/tmp/glob_test/98", "/tmp/glob_test/99"]

real    0m0,308s
user    0m0,238s
sys     0m0,075s
```
```
$ time bin/opal -r nodejs/dir -e 'p Dir.glob("/tmp/glob_test_2/*")'
["/tmp/glob_test_2/bar", "/tmp/glob_test_2/dir_not_accessible", "/tmp/glob_test_2/foo"]

real    0m0,304s
user    0m0,252s
sys     0m0,052s
```

#### After
```
$ time bin/opal -r nodejs/dir -e 'p Dir.glob("/tmp/glob_test_2/*")'
["/tmp/glob_test_2/bar", "/tmp/glob_test_2/dir_not_accessible", "/tmp/glob_test_2/foo"]

real    0m0,310s
user    0m0,242s
sys     0m0,071s
```
```
$ time bin/opal -r nodejs/dir -e 'p Dir.glob("/tmp/glob_test/*")'
["/tmp/glob_test/1", "/tmp/glob_test/10", "/tmp/glob_test/100", "/tmp/glob_test/11", "/tmp/glob_test/12", "/tmp/glob_test/13", "/tmp/glob_test/14", "/tmp/glob_test/15", "/tmp/glob_test/16", "/tmp/glob_test/17", "/tmp/glob_test/18", "/tmp/glob_test/19", "/tmp/glob_test/2", "/tmp/glob_test/20", "/tmp/glob_test/21", "/tmp/glob_test/22", "/tmp/glob_test/23", "/tmp/glob_test/24", "/tmp/glob_test/25", "/tmp/glob_test/26", "/tmp/glob_test/27", "/tmp/glob_test/28", "/tmp/glob_test/29", "/tmp/glob_test/3", "/tmp/glob_test/30", "/tmp/glob_test/31", "/tmp/glob_test/32", "/tmp/glob_test/33", "/tmp/glob_test/34", "/tmp/glob_test/35", "/tmp/glob_test/36", "/tmp/glob_test/37", "/tmp/glob_test/38", "/tmp/glob_test/39", "/tmp/glob_test/4", "/tmp/glob_test/40", "/tmp/glob_test/41", "/tmp/glob_test/42", "/tmp/glob_test/43", "/tmp/glob_test/44", "/tmp/glob_test/45", "/tmp/glob_test/46", "/tmp/glob_test/47", "/tmp/glob_test/48", "/tmp/glob_test/49", "/tmp/glob_test/5", "/tmp/glob_test/50", "/tmp/glob_test/51", "/tmp/glob_test/52", "/tmp/glob_test/53", "/tmp/glob_test/54", "/tmp/glob_test/55", "/tmp/glob_test/56", "/tmp/glob_test/57", "/tmp/glob_test/58", "/tmp/glob_test/59", "/tmp/glob_test/6", "/tmp/glob_test/60", "/tmp/glob_test/61", "/tmp/glob_test/62", "/tmp/glob_test/63", "/tmp/glob_test/64", "/tmp/glob_test/65", "/tmp/glob_test/66", "/tmp/glob_test/67", "/tmp/glob_test/68", "/tmp/glob_test/69", "/tmp/glob_test/7", "/tmp/glob_test/70", "/tmp/glob_test/71", "/tmp/glob_test/72", "/tmp/glob_test/73", "/tmp/glob_test/74", "/tmp/glob_test/75", "/tmp/glob_test/76", "/tmp/glob_test/77", "/tmp/glob_test/78", "/tmp/glob_test/79", "/tmp/glob_test/8", "/tmp/glob_test/80", "/tmp/glob_test/81", "/tmp/glob_test/82", "/tmp/glob_test/83", "/tmp/glob_test/84", "/tmp/glob_test/85", "/tmp/glob_test/86", "/tmp/glob_test/87", "/tmp/glob_test/88", "/tmp/glob_test/89", "/tmp/glob_test/9", "/tmp/glob_test/90", "/tmp/glob_test/91", "/tmp/glob_test/92", "/tmp/glob_test/93", "/tmp/glob_test/94", "/tmp/glob_test/95", "/tmp/glob_test/96", "/tmp/glob_test/97", "/tmp/glob_test/98", "/tmp/glob_test/99"]

real    0m0,309s
user    0m0,216s
sys     0m0,096s
```

takaram commented:
It still doesn't work with some patterns.

- Pattern contains no special chars like `*`, and indicates a directory, socket, fifo, device file, etc
  - `/tmp`
  - `/dev/null`
- `*` is in more than one directories
  - `/usr/*/*`

ggrossetie commented:
> Pattern contains no special chars like *, and indicates a directory, socket, fifo, device file, etc

Does Ruby consider `Dir.glob("/tmp/*")`  and `Dir.glob("/tmp")` to be equivalent?

> * is in more than one directories /usr/*/*

I did some tests with similar patterns and the results looked good. Could you please expand on that?
Also, I noticed that Glob (Node library) will return the base directory when using `**`.
I need to check if this is also what Ruby is doing.

ggrossetie commented:
@takaram Sorry for the delay. The implementation should now be closer to Ruby. I've added a few test cases.
I was surprised that `/tmp/**` is not recursive:

```console
$ ruby -v
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-linux]

$ mkdir -p /tmp/opal/foo/bar/baz

$ ruby -e 'p Dir.glob("/tmp/opal/**")'
["/tmp/opal/foo"]

$ ruby -e 'p Dir.glob("/tmp/opal/**/*")'
["/tmp/opal/foo", "/tmp/opal/foo/bar", "/tmp/opal/foo/bar/baz"]
```

ggrossetie commented:
The macOS failure does not seem to be related:

```
LoadError: dlopen(/Users/runner/work/opal/opal/vendor/bundle/ruby/3.3.0/gems/puma-6.4.2/lib/puma/puma_http11.bundle, 0x0009): Library not loaded: /Users/runner/hostedtoolcache/Ruby/3.3.4/arm64/lib/libruby.3.3.dylib (LoadError)
  Referenced from: <[9](https://github.com/opal/opal/actions/runs/10759213479/job/29835607414?pr=2667#step:14:10)B03CD19-6FFF-336E-BB32-2F9616210FEE> /Users/runner/work/opal/opal/vendor/bundle/ruby/3.3.0/gems/puma-6.4.2/lib/puma/puma_http11.bundle
  Reason: tried: '/Users/runner/hostedtoolcache/Ruby/3.3.4/arm64/lib/libruby.3.3.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/runner/hostedtoolcache/Ruby/3.3.4/arm64/lib/libruby.3.3.dylib' (no such file), '/Users/runner/hostedtoolcache/Ruby/3.3.4/arm64/lib/libruby.3.3.dylib' (no such file) - /Users/runner/work/opal/opal/vendor/bundle/ruby/3.3.0/gems/puma-6.4.2/lib/puma/puma_http11.bundle
```

hmdne commented:
Thanks @ggrossetie and @takaram :D

For merging, we would need to:
- add tests that check for all those edge cases (like block/character devices etc.). This can be done just on Unix platforms
- we have merged #2685 which can cause some conflicts

ggrossetie commented:
> we have merged https://github.com/opal/opal/pull/2685 which can cause some conflicts

I just updated the branch and it seems that we do not have any conflict 🙌🏻 

> add tests that check for all those edge cases (like block/character devices etc.). This can be done just on Unix platforms

I've added a few test cases. I can copy and adapt a few more from https://github.com/ruby/ruby/blob/master/spec/ruby/core/dir/glob_spec.rb ?

hmdne commented:
@ggrossetie It would be great if you added some test case on Unix platforms testing for those two particular cases:

```ruby
Dir.glob("/dev/null") == ["/dev/null"]
Dir.glob("/dev/*").include? "/dev/null"
```

janbiedermann commented:
Just found that node 22 has fs.globSync:
https://nodejs.org/docs/latest/api/fs.html#fsglobsyncpattern-options

janbiedermann commented:
Sure

janbiedermann commented:
#2746 provides a rather rubyish implementation of #glob

ggrossetie commented:
> https://github.com/opal/opal/pull/2746 provides a rather rubyish implementation of #glob

Looks promising, even though I don't really understand how it works 😄

henrebotha commented:
What needs to be done to move this forward?
